### PR TITLE
use task icon/color in quick pick 

### LINKED ISSUE #151758: show task icon/colors in the task quick pick
Once we have #145965, we should show them in the task quick pick like we do with the terminal profiles
Screen Shot 2022-06-10 at 8 33 14 AM

### LINKED ISSUE #151992: make task icon/color configuration more discoverable/ easier to set
on focus of a row, it would be nice if we had this pencil icon and, when pressed, it could let you set the color and icon for the task
Screen Shot 2022-06-13 at 2 47 34 PM

### PR DESCRIPTION
**Pull request by meganrogge**
Screen Shot 2022-06-13 at 4 07 29 PM
Screen Shot 2022-06-13 at 4 07 59 PM

This PR fixes #151758 
fixes #151992

To do:

- without API, we can't update the icon / color for tasks like the VS Code build task. Need to make sure the configure icon/color button doesn't show up for these since it won't do anything.

### CODE REVIEW

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts [hunk 1]
```diff
@@ -61,22 +64,36 @@ export class TaskQuickPick extends Disposable {
...
-			return task._label;
+			return TaskQuickPick.getTaskLabelWithIcon(task);
 		}
 		if (ConfiguringTask.is(task)) {
 			let label: string = task.configures.type;
 			const configures: Partial<KeyedTaskIdentifier> = Objects.deepClone(task.configures);
 			delete configures['_key'];
 			delete configures['type'];
-			Object.keys(configures).forEach(key => label += `: ${configures[key]}`);
+			Object.keys(configures).forEach(key => label += `: ${configures[key]} `);
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:74]:
```suggestion
			Object.keys(configures).forEach(key => label += `: ${configures[key]}`);
```

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts [hunk 2]
```diff
@@ -91,7 +108,7 @@ export class TaskQuickPick extends Disposable {
 	private _createTypeEntries(entries: QuickPickInput<ITaskTwoLevelQuickPickEntry>[], types: string[]) {
 		entries.push({ type: 'separator', label: nls.localize('contributedTasks', "contributed") });
 		types.forEach(type => {
-			entries.push({ label: `$(folder) ${type}`, task: type, ariaLabel: nls.localize('taskType', "All {0} tasks", type) });
+			entries.push({ label: `$(folder) ${type} `, task: type, ariaLabel: nls.localize('taskType', "All {0} tasks", type) });
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:111]:
```suggestion
			entries.push({ label: `$(folder) ${type}`, task: type, ariaLabel: nls.localize('taskType', "All {0} tasks", type) });
```

File: .vscode/tasks.json [hunk 3]
```diff
@@ -67,7 +67,9 @@
 				"kind": "build",
 				"isDefault": true
 			},
-			"problemMatcher": []
+			"problemMatcher": [],
+			"icon": "light-bulb",
+			"color": "terminal.ansiYellow"
```
Comment (meganrogge) [inline .vscode/tasks.json:72]:
```suggestion
```

File: .vscode/tasks.json [hunk 4]
```diff
@@ -67,7 +67,7 @@
 				"kind": "build",
 				"isDefault": true
 			},
-			"problemMatcher": []
+			"problemMatcher": [],
```
Comment (meganrogge) [inline .vscode/tasks.json:70]:
```suggestion
			"problemMatcher": []
```

File: .vscode/tasks.json [hunk 5]
```diff
@@ -187,7 +187,13 @@
...
 				"command": ".\\scripts\\code-server.bat"
 			},
-			"args": ["--no-launch", "--connection-token", "dev-token", "--port", "8080"],
+			"args": [
+				"--no-launch",
+				"--connection-token",
+				"dev-token",
+				"--port",
+				"8080"
+			],
```
Comment (meganrogge) [inline .vscode/tasks.json:196]:
```suggestion
```

File: .vscode/tasks.json [hunk 6]
```diff
@@ -187,7 +187,6 @@
 			"windows": {
 				"command": ".\\scripts\\code-server.bat"
 			},
```
Comment (meganrogge) [inline .vscode/tasks.json:189]:
```suggestion
			},
			"args": ["--no-launch", "--connection-token", "dev-token", "--port", "8080"],
```

**Comment by meganrogge** [comment]
Issue at the moment: 

- without API, we can't update the icon / color for tasks like the VS Code build task. Not sure how I can tell whether the edit button should be included based for a given task.

**Review by roblourens — APPROVED** [review]

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts [hunk 7]
```diff
@@ -74,9 +77,29 @@ export class TaskQuickPick extends Disposable {
...
+			document.body.appendChild(styleElement);
+		}
+	}
+
 	private _createTaskEntry(task: Task | ConfiguringTask, extraButtons: IQuickInputButton[] = []): ITaskTwoLevelQuickPickEntry {
+		const customizeIconButton = { iconClass: ThemeIcon.asClassName(Codicon.pencil), tooltip: nls.localize('setIconAndColor', "Choose color and icon") };
 		const entry: ITaskTwoLevelQuickPickEntry = { label: this._guessTaskLabel(task), description: this._taskService.getTaskDescription(task), task, detail: this._showDetail() ? task.configurationProperties.detail : undefined };
-		entry.buttons = [{ iconClass: ThemeIcon.asClassName(configureTaskIcon), tooltip: nls.localize('configureTask', "Configure Task") }, ...extraButtons];
+		entry.buttons = [];
+		if (CustomTask.is(task)) {
```
Comment (alexr00) [inline src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:97]: Why only have this for `CustomTask`? A `ContributedTask` or `ConfiguringTask` can both be modified in tasks.json to have a color and icon set.

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts [hunk 8]
```diff
@@ -211,6 +234,9 @@ export class TaskQuickPick extends Disposable {
 				if (indexToRemove >= 0) {
 					picker.items = [...picker.items.slice(0, indexToRemove), ...picker.items.slice(indexToRemove + 1)];
 				}
+			} else if (context.button.iconClass = ThemeIcon.asClassName(Codicon.pencil)) {
```
Comment (alexr00) [inline src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:237]: Single equal intended to be triple equal?

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts [hunk 9]
```diff
@@ -265,6 +291,74 @@ export class TaskQuickPick extends Disposable {
...
+			disposables.push(quickPick.onDidHide(() => r(undefined)));
+			disposables.push(quickPick.onDidAccept(() => r(quickPick.selectedItems[0])));
+		});
+		dispose(disposables);
+
+		if (result && task && typeof task !== 'string') {
+			task.configurationProperties.color = result.id;
+		}
+		document.body.removeChild(styleElement);
+		quickPick.hide();
```
Comment (alexr00) [inline src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:328]: I see in `_setIcon` you're calling `this._taskService.customize` to set the icon, which makes sense. I don't see that in `_setColor` though.

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts (hunk 7 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:97]: it won't work atm without API changes for the `ContributedTask` or `ConfiguringTask`

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts (hunk 9 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:328]: color gets set in `_setIcon`

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts (hunk 7 above)
Comment (alexr00) [reply src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:97]: I would still expect it to work for tasks in tasks.json though. A `ConfiguringTask`, is just data from tasks.json, so if the icon is defined in the tasks.json, we should be able to show it here.

Comment (Tyriar) [reply src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:97]: This was followed up on in https://github.com/microsoft/vscode/pull/152106 where it's added in `TaskDTO.to`

**Pull request merged**